packages/features/data-table/components/segment/SaveFilterSegmentButton.tsx

```diff
@@ -0,0 +1,223 @@
+import { useSession } from "next-auth/react";
+import { useEffect, useState } from "react";
+import { useForm } from "react-hook-form";
+
+import { useLocale } from "@calcom/lib/hooks/useLocale";
+import { MembershipRole, type FilterSegmentScope } from "@calcom/prisma/enums";
+import { trpc } from "@calcom/trpc/react";
+import { Button } from "@calcom/ui/components/button";
+import {
+  Dialog,
+  DialogContent,
+  DialogFooter,
+  DialogHeader,
+  DialogTrigger,
+} from "@calcom/ui/components/dialog";
+import { Form, Input, Label, Select, Switch } from "@calcom/ui/components/form";
+import { RadioGroup, RadioField } from "@calcom/ui/components/radio";
+import { showToast } from "@calcom/ui/components/toast";
+
+import { useDataTable } from "../../hooks";
+
+interface FormValues {
+  name: string;
+  scope: FilterSegmentScope;
+  teamId?: number;
+}
+
+export function SaveFilterSegmentButton() {
+  const { t } = useLocale();
+  const utils = trpc.useUtils();
+  const [isOpen, setIsOpen] = useState(false);
+  const [isTeamSegment, setIsTeamSegment] = useState(false);
+  const [selectedTeamId, setSelectedTeamId] = useState<number>();
+  const session = useSession();
+  const orgRole = session?.data?.user?.org?.role;
+  const isAdminOrOwner = orgRole === MembershipRole.OWNER || orgRole === MembershipRole.ADMIN;
+
+  const form = useForm<FormValues>({
+    defaultValues: {
+      name: "",
+      scope: "USER",
+    },
+  });
+
+  const {
+    tableIdentifier,
+    activeFilters,
+    sorting,
+    columnVisibility,
+    columnSizing,
+    selectedSegment,
+    canSaveSegment,
+    setSegmentId,
+  } = useDataTable();
+
+  const [saveMode, setSaveMode] = useState<"create" | "update">(() =>
+    selectedSegment ? "update" : "create"
+  );
+
+  const { data: teams } = trpc.viewer.teams.list.useQuery();
+
+  const { mutate: createSegment } = trpc.viewer.filterSegments.create.useMutation({
+    onSuccess: ({ id }) => {
+      utils.viewer.filterSegments.list.invalidate();
+      showToast(t("filter_segment_saved"), "success");
+      setSegmentId(id);
+      setIsOpen(false);
+    },
+    onError: () => {
+      showToast(t("error_saving_filter_segment"), "error");
+    },
+  });
+
+  const { mutate: updateSegment } = trpc.viewer.filterSegments.update.useMutation({
+    onSuccess: () => {
+      utils.viewer.filterSegments.list.invalidate();
+      showToast(t("filter_segment_updated"), "success");
+      setIsOpen(false);
+    },
+    onError: () => {
+      showToast(t("error_updating_filter_segment"), "error");
+    },
+  });
+
+  useEffect(() => {
+    if (!isOpen) {
+      setIsTeamSegment(false);
+      setSelectedTeamId(undefined);
+      setSaveMode(selectedSegment ? "update" : "create");
+      form.reset();
+    }
+  }, [isOpen, form, selectedSegment]);
```

Comment:
I think we can get rid of this effect and move it to a handleOpenChange function 

```
  const handleOpenChange = (open: boolean) => {
    if (!open) {
      // Reset form state when dialog closes
      setIsTeamSegment(false);
      setSelectedTeamId(undefined);
      setSaveMode(selectedSegment ? "update" : "create");
      form.reset();
    }
    setIsOpen(open);
  };
```

We then can pass this into the dialog `onOpenChange` props

```
    <Dialog open={isOpen} onOpenChange={handleOpenChange}>
```

Reply:
Might be worth looking at got a fair few re-renders when saving

Reply:
much nicer!